samples/300/connect_retry_with_telemetry.py

```diff
@@ -0,0 +1,171 @@
+# Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import random
+import asyncio
+from azure.iot.device.aio import IoTHubDeviceClient
+from azure.iot.device import exceptions
+import logging
+import traceback
+import os
+
+logging.basicConfig(level=logging.DEBUG, filename="debug.log")
+# For more detailed debugging can do the following
+# logging.basicConfig(
+#     level=logging.DEBUG,
+#     format="%(asctime)s %(thread)s %(funcName)s %(message)s",
+#     filename="debug1.log",
+# )
+
+# The device connection string to authenticate the device with your IoT hub.
+# Using the Azure CLI:
+# az iot hub device-identity show-connection-string --hub-name {YourIoTHubName} --device-id MyNodeDevice --output table
+CONNECTION_STRING = os.getenv("IOTHUB_DEVICE_CONNECTION_STRING")
+
+
+# The interval at which to send telemetry
+TELEMETRY_INTERVAL = 5
+# Number of times connection attempts will be made.
+RETRY_NOS = 50
+# Interval between consecutive connection attempts
+ATTEMPT_INTERVAL = 5
+connected_event = None
+disconnected_event = None
+main_event_loop = None
+
+# TODO We might have to expand our error list to be more detailed
+# And possibly have status codes that reflect what sort of a service error
+# To have a more detailed experience
+transient_errors = [
+    exceptions.OperationCancelled,
+    exceptions.OperationTimeout,
+    exceptions.ServiceError,
+    exceptions.ConnectionFailedError,
+    exceptions.ConnectionDroppedError,
+    exceptions.NoConnectionError,
+    exceptions.ClientError,  # TODO Only TLSEsxhangeError is actually worth retrying
+]
+
+
+def is_retryable(exc):
+    if type(exc) in transient_errors:
+        return True
+    else:
+        return False
+
+
+def create_client():
+    global connected_event
+    global disconnected_event
+
+    # Create a Device Client
+    device_client = IoTHubDeviceClient.create_from_connection_string(CONNECTION_STRING)
+
+    connected_event = asyncio.Event()
+    disconnected_event = asyncio.Event()
+
+    # Define a connection state change handler
+    async def handle_on_connection_state_change():
+        global main_event_loop
+        print(
+            "handle_on_connection_state_change fired. Connected status : {}".format(
+                device_client.connected
+            )
+        )
+        if device_client.connected:
+            print("Connected event is set...")
+            main_event_loop.call_soon_threadsafe(connected_event.set)
+        else:
+            print("Disconnected event is set...")
+            main_event_loop.call_soon_threadsafe(disconnected_event.set)
+
+    try:
+        # Attach the connection state handler
+        device_client.on_connection_state_change = handle_on_connection_state_change
+    except Exception:
+        # Clean up in the event of failure
+        device_client.shutdown()
+        raise
+
+    return device_client
+
+
+async def do_work(device_client):
+    global connected_event
```

Comment:
If the manual connection wasn't done each pass of the main loop as I suggested in another comment, this wouldn't need to be it's own coroutine, and could be just part of that main loop itself.

If you end up keeping this coroutine, I'm not really a fan of the name `do_work` since I feel it isn't very descriptive. Perhaps describing what the work is would be best